detector_test.go

```diff
@@ -1,108 +1,390 @@
 	spec.Run(t, "Detector", testDetector, spec.Report(report.Terminal{}))
 }
 
 func testDetector(t *testing.T, when spec.G, it spec.S) {
-	var (
-		config      *lifecycle.DetectConfig
-		platformDir string
-		tmpDir      string
-		logHandler  *memory.Handler
-	)
-
-	it.Before(func() {
-		var err error
-		tmpDir, err = ioutil.TempDir("", "lifecycle")
-		if err != nil {
-			t.Fatalf("Error: %s\n", err)
-		}
-		platformDir = filepath.Join(tmpDir, "platform")
-		appDir := filepath.Join(tmpDir, "app")
-		h.Mkdir(t, appDir, filepath.Join(platformDir, "env"))
-
-		buildpacksDir := filepath.Join("testdata", "by-id")
-
-		logHandler = memory.New()
-		config = &lifecycle.DetectConfig{
-			FullEnv:       append(os.Environ(), "ENV_TYPE=full"),
-			ClearEnv:      append(os.Environ(), "ENV_TYPE=clear"),
-			AppDir:        appDir,
-			PlatformDir:   platformDir,
-			BuildpacksDir: buildpacksDir,
-			Logger:        &log.Logger{Handler: logHandler},
-		}
-	})
+	when("#Detect", func() {
+		var (
+			mockCtrl       *gomock.Controller
+			detector       *lifecycle.Detector
+			resolver       *testmock.MockResolver
+			buildpackStore *testmock.MockBuildpackStore
+		)
+
+		it.Before(func() {
+			mockCtrl = gomock.NewController(t)
+			detector = &lifecycle.Detector{
+				Runs: &sync.Map{},
+			}
 
-	it.After(func() {
-		os.RemoveAll(tmpDir)
-	})
+			resolver = testmock.NewMockResolver(mockCtrl)
+			detector.Resolver = resolver
 
-	mkappfile := func(data string, paths ...string) {
-		t.Helper()
-		for _, p := range paths {
-			h.Mkfile(t, data, filepath.Join(config.AppDir, p))
-		}
-	}
-	toappfile := func(data string, paths ...string) {
-		t.Helper()
-		for _, p := range paths {
-			tofile(t, data, filepath.Join(config.AppDir, p))
-		}
-	}
-	rdappfile := func(path string) string {
-		t.Helper()
-		return h.Rdfile(t, filepath.Join(config.AppDir, path))
-	}
+			buildpackStore = testmock.NewMockBuildpackStore(mockCtrl)
+			detector.Store = buildpackStore
+		})
+
+		it.After(func() {
+			mockCtrl.Finish()
+		})
 
-	when("#Detect", func() {
 		it("should expand order-containing buildpack IDs", func() {
-			mkappfile("100", "detect-status")
+			bpE1 := testmock.NewMockBuildpack(mockCtrl)
+			bpA1 := testmock.NewMockBuildpack(mockCtrl)
+			bpF1 := testmock.NewMockBuildpack(mockCtrl)
+			bpC1 := testmock.NewMockBuildpack(mockCtrl)
+			bpB1 := testmock.NewMockBuildpack(mockCtrl)
+			bpG1 := testmock.NewMockBuildpack(mockCtrl)
+			bpB2 := testmock.NewMockBuildpack(mockCtrl)
+			bpC2 := testmock.NewMockBuildpack(mockCtrl)
+			bpD2 := testmock.NewMockBuildpack(mockCtrl)
+			bpD1 := testmock.NewMockBuildpack(mockCtrl)
+
+			buildpackStore.EXPECT().Lookup("E", "v1").Return(bpE1, nil)
+			bpE1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{
+				API: "0.2",
+				Order: []buildpack.Group{
+					{
+						Group: []buildpack.GroupBuildpack{
+							{ID: "A", Version: "v1"},
+							{ID: "F", Version: "v1"},
+							{ID: "B", Version: "v1"},
+						},
+					},
+				},
+			})
```

Comment:
This mirrors the structure that was implemented by our test fixtures (now deleted, since we don't have to call actual `Detect` on each buildpack toml).